storagestate/storagestate.go

```diff
@@ -0,0 +1,689 @@
+	storageState.Lock()
+	defer storageState.Unlock()
+
+	log.WithFields(log.Fields{
+		"instaceID":    instanceID,
+		"storageQuota": params.StorageQuota,
+		"stateQuota":   params.StateQuota,
+	}).Debug("Setup storage and state")
+
+	storageStateInfo, err := storageState.storage.GetStorageStateInfoByID(instanceID)
+	if err != nil {
+		if !errors.Is(err, ErrNotExist) {
+			return "", "", nil, aoserrors.Wrap(err)
+		}
+
+		if err = storageState.storage.AddStorageStateInfo(
+			instanceID, StorageStateInstanceInfo{}); err != nil {
+			return "", "", nil, aoserrors.Wrap(err)
+		}
+	}
+
+	if storagePath, err = storageState.setup(instanceID, storageState.storageDir, params.StorageQuota == 0); err != nil {
+		return "", "", nil, aoserrors.Wrap(err)
+	}
+
+	if _, ok := storageState.statesMap[params.InstanceIdent]; ok {
+		if err := storageState.stopStateWatching(params.InstanceIdent); err != nil {
+			return "", "", nil, aoserrors.Wrap(err)
+		}
+
+		delete(storageState.statesMap, params.InstanceIdent)
+	}
+
+	if statePath, err = storageState.setup(instanceID, storageState.stateDir, params.StateQuota == 0); err != nil {
+		return "", "", nil, aoserrors.Wrap(err)
+	}
+
+	if params.StateQuota != 0 {
+		defer func() {
+			if err != nil {
+				if err := storageState.stopStateWatching(params.InstanceIdent); err != nil {
+					log.Warnf("Can't stop watching state file: %v", err)
+				}
+			}
+		}()
+
+		if err = storageState.setupStateWatching(instanceID, statePath, params); err != nil {
+			return "", "", nil, aoserrors.Wrap(err)
+		}
+
+		stateChecksum = storageStateInfo.StateChecksum
+
+		storageState.statesMap[params.InstanceIdent].checksum = storageStateInfo.StateChecksum
+
+		if err = storageState.checkChecksumAndSendUpdateRequest(
+			path.Join(statePath, stateFile), params.InstanceIdent); err != nil {
+			return "", "", nil, aoserrors.Wrap(err)
+		}
+	}
+
+	if storageStateInfo.StateQuota != params.StateQuota || storageStateInfo.StorageQuota != params.StorageQuota {
+		if err = storageState.setQuotasFS(params); err != nil {
+			return "", "", nil, aoserrors.Wrap(err)
+		}
+
+		if err = storageState.storage.SetStorageStateQuotasByID(
+			instanceID, params.StorageQuota, params.StateQuota); err != nil {
+			return "", "", nil, aoserrors.Wrap(err)
+		}
+	}
+
+	return storagePath, statePath, stateChecksum, nil
+}
+
+// Cleanup clean storagestate instance.
+func (storageState *StorageState) Cleanup(instanceID string) error {
+	storageState.Lock()
+	defer storageState.Unlock()
+
+	log.WithField("instaceID", instanceID).Debug("Clean storage and state")
+
+	for instanceIdent, state := range storageState.statesMap {
+		if state.instanceID == instanceID {
+			if err := storageState.stopStateWatching(instanceIdent); err != nil {
+				return aoserrors.Wrap(err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// Remove remove storagestate dirs.
+func (storageState *StorageState) Remove(instanceID string) error {
```

Comment:
should you stop watching in case of removal?

Reply:
Done